ghcide/src/Development/IDE/Core/Shake.hs

```diff
@@ -1232,9 +1224,10 @@ defineEarlyCutoff' doDiagnostics cmp key file mbOld mode action = do
                         _                                     -> False
                 return $ RunResult
                     (if eq then ChangedRecomputeSame else ChangedRecomputeDiff)
-                    (encodeShakeValue bs) $
-                    A res
-        liftIO $ atomicallyNamed "define - dirtyKeys" $ modifyTVar' dirtyKeys (deleteKeySet $ toKey key file)
+                    (encodeShakeValue bs)
+                    (A res) $ do
+                        setValues state key file res (Vector.fromList diags)
+                        modifyTVar' dirtyKeys (deleteKeySet $ toKey key file)
```

Comment:
Oh hmm, we still have to modify the dirty keys here? How awkward. Why do we need to do that specially here?

Reply:
Actually, this makes me think I don't understand what's going on. If it's safe to modify the TVar here (presumably because we're doing it in STM so we can put it in a transaction appropriately?), then why can't we do that elsewhere (i.e. mark things dirty in STM as needed).

Reply:
We are putting it into the hook to run after all computation is done and put it to run with marking the key clean in https://github.com/haskell/haskell-language-server/pull/4190#discussion_r1593175057

yep it should need an explicit note to demonstrate the reason

Reply:
So they get in sync. Pretty nasty.

Reply:
done